classes/manager.php

```diff
@@ -163,12 +168,190 @@ public static function get_reasons(float $duration): int {
         if (isset($SESSION->toolexcimerflameall)) {
             $reason |= self::REASON_FLAMEALL;
         }
-        if (($duration * 1000) >= (int) get_config('tool_excimer', 'trigger_ms')) {
+
+        if (($duration * 1000) >= self::get_min_duration_for_reason_slow()) {
             $reason |= self::REASON_AUTO;
         }
         return $reason;
     }
 
+    /**
+     * Returns the minimum duration for profiles matching this reason and page/request.
+     *
+     * @param  int $reason - the profile type or REASON_*
+     * @return float duration (in milliseconds) of the fastest profile for a given reason and request/page.
+     */
+    public static function get_min_duration_for_request_and_reason(string $request, int $reason): float {
+        global $DB;
+
+        $reasonstr = self::REASON_STR_MAP[$reason];
+        $pagequota = (int) get_config('tool_excimer', "num_' . $reasonstr . '_by_page");
+
+        // Grab the fastest profile for this page/request, and use that as
+        // the lower boundary for any new profiles of this page/request.
+        $cachekey = "profile_type_$reason" . "_page_$request" . '_min_duration_ms';
+        $cache = \cache::make('tool_excimer', 'timings');
+        $result = $cache->get($cachekey);
+        if ($result === false) {
+            // NOTE: Opting to query this way instead of using MIN due to
+            // the fact valid profiles will be added and the limits will be
+            // breached for 'some time'. This will keep the constraints as
+            // correct as possible.
+            $reasons = $DB->sql_bitand('reason', $reason);
+            $sql = "SELECT duration as min_duration
+                      FROM {tool_excimer_profiles}
+                     WHERE $reasons != ?
+                           AND request = ?
+                  ORDER BY duration DESC
+                     ";
+            $resultset = $DB->get_records_sql($sql, [
+                self::REASON_NONE,
+                $request,
+            ], 0, $pagequota);
```

Comment:
I think this can be simpler and we only grab the Nth record, and if there is < N records then we won't get a record and we can cache 0 instead. By not ever storing anything other than the exact Nth record we can make the logic simpler in get_min_duration_for_reason_slow

Reply:
Very nice catch here. Yes we can ignore the quota lookups with this as well.